File: samples/java/SupportChat/Server/Support/src/main/java/systems/zlink/samples/supportchat/server/support/spots/entryspot/handlers/JoinConversationActorHandler.java
package systems.zlink.samples.supportchat.server.support.spots.entryspot.handlers;

import systems.zlink.framework.ZLinkMessageContext;
import systems.zlink.framework.spots.ZLinkEntrySpotActorRequestHandler;
import systems.zlink.samples.supportchat.server.configuration.SampleNames;
import systems.zlink.samples.supportchat.server.support.actors.SupportUserActor;
import systems.zlink.samples.supportchat.server.support.spots.entryspot.SupportEntrySpot;
import systems.zlink.samples.supportchat.shared.contracts.Messages;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

public final class JoinConversationActorHandler
        implements ZLinkEntrySpotActorRequestHandler<
                SupportEntrySpot,
                SupportUserActor,
                Messages.JoinConversationReq,
                Messages.JoinConversationRes> {
    @Override
    public CompletionStage<Messages.JoinConversationRes> handle(
            SupportEntrySpot spot,
            SupportUserActor actor,
            ZLinkMessageContext context,
            Messages.JoinConversationReq request) {
        if (!SampleNames.Roles.Agent.equals(actor.role())) {
            throw new IllegalStateException(
                    "Only agent conversation actors can join through the Entry Spot");
        }
        String conversationId = request.conversationId();
        if (conversationId == null || conversationId.isBlank()) {
            throw new IllegalStateException("Conversation Join is missing conversationId");
        }
        Messages.JoinConversationRes scheduled =
                actor.scheduleConversationJoin(
                        conversationId,
                        "",
                        new Messages.JoinConversationReq(
                                conversationId,
                                actor.participantId(),
                                actor.role(),
                                actor.displayName()));
        return CompletableFuture.completedFuture(scheduled);
    }
}

File: samples/kotlin/SupportChat/Server/Support/src/main/kotlin/systems/zlink/samples/kotlin/supportchat/server/support/infrastructure/zlink/spots/entryspot/handlers/JoinConversationActorHandler.kt
package systems.zlink.samples.kotlin.supportchat.server.support.infrastructure.zlink.spots.entryspot.handlers

import systems.zlink.framework.ZLinkMessageContext
import systems.zlink.framework.kotlin.ZLinkSuspendingEntrySpotActorRequestHandler
import systems.zlink.samples.kotlin.supportchat.server.configuration.SampleNames
import systems.zlink.samples.kotlin.supportchat.server.configuration.SupportChatRoles
import systems.zlink.samples.kotlin.supportchat.server.support.infrastructure.zlink.actors.SupportUserActor
import systems.zlink.samples.kotlin.supportchat.server.support.infrastructure.zlink.spots.entryspot.SupportEntrySpot
import systems.zlink.samples.kotlin.supportchat.shared.contracts.JoinConversationReq
import systems.zlink.samples.kotlin.supportchat.shared.contracts.JoinConversationRes

class JoinConversationActorHandler :
    ZLinkSuspendingEntrySpotActorRequestHandler<
        SupportEntrySpot,
        SupportUserActor,
        JoinConversationReq,
        JoinConversationRes,
    > {
    override suspend fun handle(
        entrySpot: SupportEntrySpot,
        actor: SupportUserActor,
        context: ZLinkMessageContext,
        request: JoinConversationReq,
    ): JoinConversationRes {
        if (actor.role != SupportChatRoles.Agent) {
            throw IllegalStateException(
                "Only agent conversation actors can join through the Entry Spot"
            )
        }
        val conversationId =
            context.metadata()[SampleNames.ConversationIdMetadataKey]?.takeIf(String::isNotBlank)
                ?: throw IllegalStateException(
                    "Conversation Join is missing the conversation ID metadata"
                )
        return actor.scheduleConversationJoin(
            conversationId,
            "",
            JoinConversationReq(actor.participantId, actor.role, actor.displayName),
        )
    }
}
